mlx_graphs/nn/conv/rel_conv.py

```diff
@@ -0,0 +1,336 @@
+
+        Returns:
+            The computed node embeddings
+        """
+        self.input_dims = len(node_features.shape)
+
+        batch_size = node_features.shape[0] if self.input_dims == 3 else 1
+        if size is None:
+            num_nodes = (
+                node_features.shape[0]
+                if self.input_dims == 2
+                else node_features.shape[1]
+            )
+            size = (num_nodes, num_nodes)
+
+        # input: (bs, num_nodes, dim)
+        if self.dependent:
+            assert query is not None, "query must be supplied when dependent=True"
+            assert (
+                self.input_dims == 3
+            ), "expected input shape is [batch_size, num_nodes, dim]"
+            # relation features as a projection of input "query" (relation) embeddings
+            relation = self.relation_linear(query).reshape(
+                batch_size, self.num_relation, self.input_dim
+            )
+        else:
+            # relation features as an embedding matrix unique to each layer
+            # relation: (batch_size, num_relation, dim)
+            relation = mx.repeat(self.relation.weight[None, :], batch_size, axis=0)
+
+        if edge_weight is None:
+            edge_weight = mx.ones(len(edge_type))
+
+        # since mlx_graphs gathers always along dimension 0 (num_nodes are rows)
+        # we have to reshape input features accordingly
+        if self.input_dims == 3:
+            node_features = node_features.transpose(1, 0, 2)
+            boundary = boundary.transpose(1, 0, 2)
+
+        # note that we send the initial boundary condition (node states at layer0)
+        # to the message passing
+        # correspond to Eq.6 on p5 in https://arxiv.org/pdf/2106.06935.pdf
+        output = self.propagate(
+            node_features=node_features,
+            edge_index=edge_index,
+            message_kwargs=dict(
+                relation=relation, boundary=boundary, edge_type=edge_type
+            ),
+            aggregate_kwargs=dict(edge_weight=edge_weight, dim_size=size),
+            update_kwargs=dict(old=node_features),
+        )
+        return output
+
+    def message(
+        self,
+        src_features: mx.array,
+        dst_features: mx.array,
+        relation: mx.array,
+        boundary: mx.array,
+        edge_type: mx.array,
+    ) -> mx.array:
+        # extracting relation features
+        relation_j = relation[:, edge_type]
+
+        if self.input_dims == 3:
+            relation_j = relation_j.transpose(1, 0, 2)
+        else:
+            relation_j = relation_j.squeeze(0)
+
+        if self.message_func == "transe":
+            message = src_features + relation_j
+        elif self.message_func == "distmult":
+            message = src_features * relation_j
+        elif self.message_func == "rotate":
+            x_j_re, x_j_im = src_features.chunk(2, dim=-1)
+            r_j_re, r_j_im = relation_j.chunk(2, dim=-1)
```

Comment:
@migalkin this gives me an error - I'm not sure there's a `chunk` method implemented for mlx arrays